### PR TITLE
Validate only .po files that are not product of a merge

### PR DESCRIPTION
**Pull request by baxeico**
See https://openedx.atlassian.net/browse/LOC-96. This PR is covered by the OpenCraft contributor agreement. @antoviaque

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @baxeico! I've created [OSPR-1278](https://openedx.atlassian.net/browse/OSPR-1278) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](http://open.edx.org/sites/default/files/wysiwyg/individual-contributor-agreement.pdf)     or indicated your institutional affiliation.     If you like, you can add yourself to the [AUTHORS](https://github.com/edx/i18n-tools/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

**Comment by jbzdak** [comment]
Hi @baxeico, I'm an OpenCraft engineer assigned to do a first round of review of this PR.  

I have tried testing this PR, and it it looks like it filters out wrong files. The LOC-96 description states that we shouldn't find errors in `django-partial.po` as if there are errors in `django-partial.po`, `django.po` will have errors also. 

This PR disables testing for `django.po` and `djangojs.po` files (as `config.CONFIGURATION.generate_merge.keys()` returns `['djangojs.po', 'django.po']`). 

Could you please fix that and ping me?

File: i18n/validate.py [hunk 1]
```diff
@@ -26,8 +26,12 @@ def validate_po_files(root, report_empty=False):
...
         for name in filenames:
             __, ext = os.path.splitext(name)
-            if ext.lower() == '.po':
-                filename = os.path.join(dirpath, name)
+            filename = os.path.join(dirpath, name)
+
+            # validate only .po files that are not product of a merge (see generate.py)
+            # if django-partial.po has a problem, then django.po will also, so don't report it.
+            if (ext.lower() == '.po' and
+                    os.path.basename(filename) not in config.CONFIGURATION.generate_merge.keys()):
```
Comment (jbzdak) [inline i18n/validate.py]: Nitpick: this condition is quite long, could you make it shorter and more readable, eg. by moving `config.CONFIGURATION.generate_merge.keys()` to a local variable?

**Comment by baxeico** [comment]
Hi @jbzdak , thank you very much for your review! I studied a bit the i18n workflow and AFAIU there are some smaller .po files (django-partial.po, mako.po, ...) that are sent to translators on Transifex and then combined into bigger files (django.po, djangojs.po) to be compiled and used in the project.

The problem was that before my patch both "partial" files and "merged" files was validated, resulting in a duplication of errors if something was wrong. My patch leaves validation only on partial files, avoiding a redundant validation on merged files. I think that reporting the problems on smaller files is more useful, because those are the files that are sent to Transifex (see https://github.com/edx/edx-platform/blob/master/.tx/config).

Also the title of the ticket is "Don't validate files that are combinations of other validated files", so this suggested my approach of skipping the products of the merge (`'djangojs.po', 'django.po'`) in the validation of files.

Please let me know what you think, and thank you again for taking the time to review my commit.

File: i18n/validate.py (hunk 1 above)
Comment (baxeico) [inline i18n/validate.py]: Ok, done here https://github.com/edx/i18n-tools/pull/41/commits/99b5c2ef2cf40ba67a30615474ae855c0ffecb22

File: i18n/validate.py [hunk 2]
```diff
@@ -26,8 +26,13 @@ def validate_po_files(root, report_empty=False):
...
         for name in filenames:
             __, ext = os.path.splitext(name)
-            if ext.lower() == '.po':
-                filename = os.path.join(dirpath, name)
+            filename = os.path.join(dirpath, name)
+
+            # validate only .po files that are not product of a merge (see generate.py)
+            # if django-partial.po has a problem, then django.po will also, so don't report it.
+            merged_files = config.CONFIGURATION.generate_merge.keys()
+            if (ext.lower() == '.po' and
```
Comment (jbzdak) [inline i18n/validate.py]: Nitpick: Now this will fit into a single line :)

**Comment by jbzdak** [comment]
@baxeico Thanks for quick response. I have done some extra digging around the issue, and it turns out you were right all along, `django.po` are indeed automatically merged from other files --- which makes validation of these other, smaller, files more sensible thing to do. 

I have verified this works as intended! And after fixing last nitpick :+1: .

File: i18n/validate.py (hunk 2 above)
Comment (baxeico) [inline i18n/validate.py]: Of course. Done!

**Comment by baxeico** [comment]
Thank you @jbzdak ! I hope that @nedbat will find some time to have a look at this PR, and eventually merge it, if it's ok also for him.

File: i18n/validate.py [hunk 3]
```diff
@@ -26,8 +26,12 @@ def validate_po_files(root, report_empty=False):
     for dirpath, __, filenames in os.walk(root):
         for name in filenames:
             __, ext = os.path.splitext(name)
-            if ext.lower() == '.po':
-                filename = os.path.join(dirpath, name)
+            filename = os.path.join(dirpath, name)
+
+            # validate only .po files that are not product of a merge (see generate.py)
+            # if django-partial.po has a problem, then django.po will also, so don't report it.
+            merged_files = config.CONFIGURATION.generate_merge.keys()
```
Comment (nedbat) [inline i18n/validate.py]: This variable doesn't change, so it can be moved up above the double for-loops.

File: i18n/validate.py [hunk 4]
```diff
@@ -26,8 +26,12 @@ def validate_po_files(root, report_empty=False):
     for dirpath, __, filenames in os.walk(root):
         for name in filenames:
             __, ext = os.path.splitext(name)
-            if ext.lower() == '.po':
-                filename = os.path.join(dirpath, name)
+            filename = os.path.join(dirpath, name)
+
+            # validate only .po files that are not product of a merge (see generate.py)
```
Comment (nedbat) [inline i18n/validate.py]: Sorry to be picky, but would you mind using English sentence style for comments?  Capitalize the first letter, and use periods at the end of sentences.  It makes the comments easier to read and understand.

File: i18n/validate.py [hunk 5]
```diff
@@ -171,6 +175,10 @@ def report_problems(filename, problems):
 
 
 class Validate(Runner):
+    """
+    Implements Runner interface
+    """
```
Comment (nedbat) [inline i18n/validate.py]: Thanks for ensuring the class has a docstring, but this docstring doesn't explain anything about the class.  Let's change it to:  "The `validate` command for checking that .po files have no errors."

**Comment by nedbat** [comment]
@baxeico Thanks for doing this work, it's been a long-standing aggravation! :)

Just a few tiny changes, then squash your commits, and we are good to merge!

**Comment by baxeico** [comment]
Thank you @nedbat ! I made the suggested changes and I squashed my commits in a single one.

**Comment by nedbat** [comment]
Thanks!

**Pull request merged**